### PR TITLE
Scala 2.13 support.

### PR DESCRIPTION
**Pull request by dotbg**
xgboost4j support of scala 2.13 #8650  and #6596

### CODE REVIEW

**Comment by trivialfis** [comment]
Thank you for the work on updating the scala version. I will defer the review to @wbo4958 as I'm not familiar with the ecosystem.

**Review by eejbyfeldt — COMMENTED** [review]
I left a review since I also looked into adding support for scala 2.13. But since I not a maintainer for this project I guess you have to take them for what they are.

Also if I run a compile on this branch
```
$ mvn compile -Pdefault,scala-2.13
...
[ERROR] Failed to execute goal on project xgboost4j-example_2.13: Could not resolve dependencies for project ml.dmlc:xgboost4j-example_2.13:jar:2.0.0-SNAPSHOT: Could not find artifact ml.dmlc:xgboost4j_2.12:jar:2.0.0-SNAPSHOT in central_maven (https://repo1.maven.org/maven2) -> [Help 1]
```
it fails looking for some `2.12`  version of the `xgboost4j` jar. Not sure if this just for me or if there is something more that was missed.

**Comment by trivialfis** [comment]
> But since I not a maintainer for this project I guess you have to take them for what they are.

You are more than welcome to contribute through any means. Also, we are looking for new maintainers for jvm packages and R.

File: jvm-packages/pom.xml
Comment (dotbg) [reply jvm-packages/pom.xml:40]: thanks. actually, this is a leftover.

File: jvm-packages/README.md
Comment (dotbg) [reply jvm-packages/README.md:48]: @eejbyfeldt unfortunately, you have to do `mvn install -Pdefault,scala-2.13` instead of simply `mvn compile` :(

File: jvm-packages/pom.xml
Comment (dotbg) [reply jvm-packages/pom.xml:80]: I wanted to be explicit, but you are right, it is better to remove this.

File: jvm-packages/xgboost4j-spark/pom.xml
Comment (dotbg) [reply jvm-packages/xgboost4j-spark/pom.xml:12]: It is safe for now. Sadly, making the script as spark adds more complexity than just using profiles. 
I've reused the approach discussed in the https://github.com/davidB/scala-maven-plugin/issues/177 
Spark approach originates in 2015 and maven changed since. I can make a script in place of maven profiles, if necessary.

**Comment by dotbg** [comment]
> . Also, we are looking for new maintainers for jvm packages and R.

@trivialfis  I have few things I'd like to do in the `jvm-packages` and have some spare time for that, so I can lend a hand for a while.

**Comment by trivialfis** [comment]
Thank you for sharing your interest! Let's work through those items first.

**Review by trivialfis — COMMENTED** [review]
Few questions, still trying to make time to learn about the spark ecosystem. ;-)

Otherwise looks fine to me. cc @wbo4958 .

File: jvm-packages/pom.xml
Comment (dotbg) [reply jvm-packages/pom.xml:8]: No, it is not changing the jar name, as the relevant project is not published as jar. 
```
    <packaging>pom</packaging>
```
Pom-packaging is a way to specify the container of submodules. It does not lead to a jar.
`jvm-packages` only produce jars for `xgboost4j-spark`, `xgboost4j`, `xgboost4j-flink`, `xgboost4j-example` and the *gpu-enabled* modules

File: jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/EvalTrait.scala
Comment (dotbg) [reply jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/EvalTrait.scala:40]: Yes, scala 2.13 is more strict in this case. The parent interface is defined as follows

```
public interface IEvaluation extends Serializable {
  /**
   * get evaluate metric
   *
   * @return evalMetric
   */
  String getMetric();

  /**
   * evaluate with predicts and data
   *
   * @param predicts predictions as array
   * @param dmat     data matrix to evaluate
   * @return result of the metric
   */
  float eval(float[][] predicts, DMatrix dmat);
}
```
this forces all methods we implement to be public. This was changed from the way it was handled in 2.12

File: jvm-packages/README.md
Comment (dotbg) [reply jvm-packages/README.md:124]: sadly, the https://github.com/NVIDIA/spark-rapids/issues/1525 is a blocker for the GPU-enabled modules. I do not think this has to be a showstopper for 2.13.

**Comment by wbo4958** [comment]
great job. I will review it. Sorry for that

File: jvm-packages/README.md
Comment (wbo4958) [inline jvm-packages/README.md:50]: Any reason to delete this line?

File: jvm-packages/README.md
Comment (wbo4958) [reply jvm-packages/README.md:124]: Seems yes.

File: jvm-packages/README.md
Comment (wbo4958) [reply jvm-packages/README.md:50]: ignore the above comment.

**Comment by wbo4958** [comment]
Really great job. LGTM assuming the CI can pass

**Comment by dotbg** [comment]
Thanks a lot @wbo4958 I've updated the `xgboost4j-tester` so that it checks both 2.12 and 2.13.

**Comment by dotbg** [comment]
I am not sure whether I can do anything for docs and python tests. 
@trivialfis , @wbo4958  is it a known issue?

**Comment by dotbg** [comment]
Resolved conflicts.

File: jvm-packages/pom.xml
Comment (trivialfis) [reply jvm-packages/pom.xml:8]: Thank you for the information!

**Review by trivialfis — APPROVED** [review]

**Comment by trivialfis** [comment]
Will merge once all tests are green. Thank you for the work on supporting the newer Scala!

**Comment by dotbg** [comment]
@trivialfis had to modify the docker image used for the integration tests. Should be green now.

**Comment by dotbg** [comment]
updated the logic, used to build images for the integration tests + rebased

**Comment by wbo4958** [comment]
> Thank you for the update! Are you able to reproduce this error: https://buildkite.com/xgboost/xgboost-ci-multi-gpu/builds/2184#01881fd5-fa29-4540-92f4-eb8861dab13b ?
> 
> Feel free to let me know if there's anything I can help.

Oops, spark-rapids 23.04.0 doesn't support spark3.4. I will fix it.

**Comment by dotbg** [comment]
@wbo4958  thanks a lot! Reproducing some things locally requires patching, so it takes more time than just writing code. 
I am not sure whether some of the changes in CI scripts are worth making a PR with these.

**Comment by trivialfis** [comment]
Restarted the CI due to a network error. Maven seems quite unstable on the CI recently.

File: jvm-packages/xgboost4j-example/src/test/resources/iris.csv
Comment (trivialfis) [inline jvm-packages/xgboost4j-example/src/test/resources/iris.csv:1]: Please don't copy data in the git repository.

Comment (dotbg) [reply jvm-packages/xgboost4j-example/src/test/resources/iris.csv:1]: @trivialfis  this file was not present in this repository at all. It is redownloaded every time for the integration tests. However, reusing this approach is not suitable for unittests :( I could consider rewriting examples to reuse existing datasets, but this would mean quite some rework :( 
I want to have the unittests present as testing that examples still work would require bildkite approval otherwise.

Comment (trivialfis) [reply jvm-packages/xgboost4j-example/src/test/resources/iris.csv:1]: Hi, thank you for sharing the consideration. I think we will have to leave the unittest out for this PR or write the unittest in a different way, like using existing datasets in XGBoost or generating a synthetic dataset.

Comment (dotbg) [reply jvm-packages/xgboost4j-example/src/test/resources/iris.csv:1]: Hm, ok. I'll make sure that the unittests generate synthetic data which follows the iris dataset structure.

Comment (trivialfis) [reply jvm-packages/xgboost4j-example/src/test/resources/iris.csv:1]: Perhaps the example can use synthetic data instead?

Comment (dotbg) [reply jvm-packages/xgboost4j-example/src/test/resources/iris.csv]: @trivialfis this would be quite a big change for this pull request. I'd like these to be done separately

**Review by trivialfis — APPROVED** [review]
Overall looks good to me, one minor issue in the comments.

Thank you for the nice work on not only updating the Scala version but also handling the tests!

File: jvm-packages/xgboost4j-example/src/test/resources/iris.csv
Comment (trivialfis) [reply jvm-packages/xgboost4j-example/src/test/resources/iris.csv]: No need for now.

File: jvm-packages/xgboost4j-example/src/test/scala/ml/dmlc/xgboost4j/scala/example/spark/SparkExamplesTest.scala
Comment (trivialfis) [inline jvm-packages/xgboost4j-example/src/test/scala/ml/dmlc/xgboost4j/scala/example/spark/SparkExamplesTest.scala:67]: Please remove the `println`. I think I have this comment in the last review but somehow GitHub didn't show it. ;-(

Comment (dotbg) [reply jvm-packages/xgboost4j-example/src/test/scala/ml/dmlc/xgboost4j/scala/example/spark/SparkExamplesTest.scala]: oh, indeed. removed.

**Comment by dotbg** [comment]
@trivialfis Thanks a lot !

**Pull request merged**
